plat/xen/x86/pvh.c

```diff
@@ -0,0 +1,176 @@
+#include <xen-x86/pvh.h>
+#include <uk/essentials.h>
+#include <uk/plat/common/lcpu.h>
+#include <uk/plat/common/bootinfo.h>
+#include <uk/plat/bootstrap.h>
+#include <uk/reloc.h>
+
+#include <stdint.h>
+#include <x86/cpu.h>
+#include <xen/xen.h>
+#include <xen/arch-x86/cpuid.h>
+#include <xen-x86/hypercall64.h>
+#include <x86/traps.h>
+#include <common/events.h>
+#include <common/hypervisor.h>
+
+shared_info_t *HYPERVISOR_shared_info;
+void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi);
+
+static void hpc_init(void)
+{
+    uint32_t eax, ebx, ecx, edx, base;
+
+    for ( base = XEN_CPUID_FIRST_LEAF;
+          base < XEN_CPUID_FIRST_LEAF + 0x10000; base += 0x100 )
+    {
+        cpuid(base, 0, &eax, &ebx, &ecx, &edx);
+
+        if ( (ebx == XEN_CPUID_SIGNATURE_EBX) &&
+             (ecx == XEN_CPUID_SIGNATURE_ECX) &&
+             (edx == XEN_CPUID_SIGNATURE_EDX) &&
+             ((eax - base) >= 2) )
+            break;
+    }
+
+    cpuid(base + 2, 0, &eax, &ebx, &ecx, &edx);
+    wrmsrl(ebx, (unsigned long)&hypercall_page);
+    barrier();
+}
+
+static void pvh_init_cmdline(struct ukplat_bootinfo *bi,
+			     struct pvh_start_info *pi)
+{
+	struct ukplat_memregion_desc mrd = {0};
+	int rc;
+
+	bi->cmdline = pi->cmdline_paddr;
+	if (unlikely(pi->cmdline_paddr == NULL))
+		return;
+
+	bi->cmdline_len = strlen((char *) pi->cmdline_paddr);
+	if (bi->cmdline_len == 0)
+		return;
+
+	mrd.type = UKPLAT_MEMRT_RESERVED;
+	mrd.flags = UKPLAT_MEMRF_READ;
+	mrd.pbase = PAGE_ALIGN_DOWN(pi->cmdline_paddr);
+	mrd.vbase = mrd.pbase;
+	mrd.len = PAGE_ALIGN_UP(bi->cmdline_len);
+	mrd.pg_count = PAGE_COUNT(mrd.pg_off + mrd.len);
```

Comment:
I don't exactly know how Unikraft manages this, but the cmdline part can be reclaimed once it's not useful anymore so we avoid losing the page.